server: remove global defaultZoneConfig

jeffrey-xiao opened the pull request:
Previously the default zone config was stored as a global and code in various layers would use `config.DefaultZoneConfig()` directly. It was difficult for parallel tests to modify the global and restore it
correctly.

The default zone config can be overridden in tests by setting the `DefaultZoneConfigOverride` testing knob.

Fixes #37054

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=37434) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cockroachdb/cockroach?pullRequest=37434) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cockroachdb/cockroach?pullRequest=37434) it.</sub>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37434)
<!-- Reviewable:end -->

ajwerner reviewed (changes requested):
It'd be nice to also get rid `defaultSystemZoneConfig` and its parallel logic while you're here. I'll come back and check out the testing updates after.

> Reviewed 1 of 9 files at r1, 2 of 63 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-:-LeXRMIYBV9fTrcYB7vu:b-8tdpkd)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @jeffrey-xiao, and @nvanbenschoten)

___
*[pkg/config/system.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXNtPj9wndemJsZ9VO:-LeXNtPkCHjgXHz0KibV:b-9kb46y) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfea24909bdebf1fb9462f2c06b3e6a9d819c53/pkg/config/system.go#L74)):*
> ```Go
> type SystemConfig struct {
> 	SystemConfigEntries
> 	DefaultZoneConfig *ZoneConfig
> ```

bit of a nit: why a pointer here?
___
*[pkg/config/zone.go, line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXQttJ0aucOOlbCFb5:-LeXQttJ0aucOOlbCFb6:bitxg9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/config/zone.go#L271)):*
> ```Go
> const minRangeMaxBytes = 64 << 10 // 64 KB
> 
> // defaultSystemZoneConfig is the default zone configuration used when no custom
> ```

This feels like it deserves the same treatment as the other zone config. 
___
*[pkg/config/zone.go, line 336 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXQc2f-U2tx0MGVQZQ:-LeXQc2gCcDeUSPky-x_:b-hjqgfg) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/config/zone.go#L336)):*
> ```Go
> // config has been specified.
> func DefaultZoneConfigRef() *ZoneConfig {
> 	return &ZoneConfig{
> ```

Make this 
```
zc := DefaultZoneConfig()
return &zc
```


<!-- Sent from Reviewable.io -->

nvb reviewed (changes requested):
This is looking good. Nice job!

> Reviewed 4 of 9 files at r1, 61 of 63 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-:-LeXs-Z50EFn9zXY9vd0:b-ft0gji)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jeffrey-xiao)

___
*[pkg/config/system.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXNtPj9wndemJsZ9VO:-LeXlO3WE9eiAJiapPtx:b1d0fp7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfea24909bdebf1fb9462f2c06b3e6a9d819c53/pkg/config/system.go#L74)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

bit of a nit: why a pointer here?
</blockquote></details>

ZoneConfig is 144 bytes large, so I think it's probably a good idea not to pass it around by value. If only we had `const &` in go.
___
*[pkg/config/zone.go, line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXQttJ0aucOOlbCFb5:-LeXlmIBEo48otj_BGon:b-c35djy) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/config/zone.go#L271)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This feels like it deserves the same treatment as the other zone config. 
</blockquote></details>

@jeffrey-xiao and I talked about this and we decided to fix it in the same way in a follow-up PR. That way we can be sure we all agree on the approach for this one first. 
___
*[pkg/config/zone.go, line 356 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXn0LY8irrfrxZW1nX:-LeXn0LY8irrfrxZW1nY:b2fyf5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/9485c3d76adcb0a6a0f1ef1750b8ef0ba1ca2e63/pkg/config/zone.go#L356)):*
> ```Go
> // TestingSetDefaultZoneConfig is a testing-only function that changes the
> // default zone config and returns a function that reverts the change.
> func TestingSetDefaultZoneConfig(cfg ZoneConfig) func() {
> ```

:fireworks: 
___
*[pkg/server/config.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXoJPt6GJPfK21mYD3:-LeXoJPu1FcrOP54g-Jr:bcuw8iw) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/server/config.go#L124)):*
> ```Go
> 	// Embed the base context.
> 	*base.Config
> 	DefaultZoneConfig *config.ZoneConfig
> ```

Give this field a comment. Can it be nil? What does it mean to set it to a non-default value?
___
*[pkg/server/config.go, line 124 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXpFzJDqh_6BW0sypX:-LeXpFzJDqh_6BW0sypY:b-36i7p2) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/server/config.go#L124)):*
> ```Go
> 	// Embed the base context.
> 	*base.Config
> 	DefaultZoneConfig *config.ZoneConfig
> ```

Nit: I'd move this down right above the `Locality` field.
___
*[pkg/server/server_update.go, line 35 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXpRCXC8c-W2cOzIpJ:-LeXpRCXC8c-W2cOzIpK:b14vzei) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/server/server_update.go#L35)):*
> ```Go
> 	// automatic version upgrade mechanism.
> 	DisableAutomaticVersionUpgrade int32 // accessed atomically
> 	DefaultZoneConfigOverride      *config.ZoneConfig
> ```

Give this a comment explaining what it means for it to be set.
___
*[pkg/sql/planner.go, line 224 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXpjGvC6UpVfqtu03d:-LeXpjGvC6UpVfqtu03e:bmem4xb) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/sql/planner.go#L224)):*
> ```Go
> 	}
> 	systemCfg := config.NewSystemConfig(execCfg.DefaultZoneConfig)
> 	systemCfg.DefaultZoneConfig = execCfg.DefaultZoneConfig
> ```

Is this line needed?
___
*[pkg/sql/set_zone_config.go, line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXpxED0EJyuDg8u1k2:-LeXpxED0EJyuDg8u1k3:b-x0rdff) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/sql/set_zone_config.go#L314)):*
> ```Go
> 		// ranges (liveness, meta, etc.), and did not have a zone config
> 		// already.
> 		completeZone = protoutil.Clone(params.extendedEvalCtx.ExecCfg.DefaultZoneConfig).(*config.ZoneConfig)
> ```

Why do we need the clone here and below?
___
*[pkg/sql/show_trace_replica_test.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXq3B5-9iStjYfK8Xy:-LeXq3B5-9iStjYfK8Xz:b-8hsxgx) ([raw file](https://github.com/cockroachdb/cockroach/blob/9485c3d76adcb0a6a0f1ef1750b8ef0ba1ca2e63/pkg/sql/show_trace_replica_test.go#L44)):*
> ```Go
> 	cfg.NumReplicas = proto.Int32(1)
> 	defer config.TestingSetDefaultZoneConfig(cfg)()
> 	defer config.TestingSetDefaultSystemZoneConfig(cfg)()
> ```

We're able to get rid of this one already?
___
*[pkg/sql/zone_config_test.go, line 340 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXqAYQCW1Ifsi9ZqTc:-LeXqAYQCW1Ifsi9ZqTd:bototn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/sql/zone_config_test.go#L340)):*
> ```Go
> 	}
> 
> 	cfg := config.DefaultSystemZoneConfig()
> ```

While you're here, `s/cfg/defaultSystemZoneConfig/`.
___
*[pkg/sql/zone_test.go, line 268 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXqLRT7uuuiLyW5h9f:-LeXqLRT7uuuiLyW5h9g:b-7yaemt) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/sql/zone_test.go#L268)):*
> ```Go
> 	newReplicationFactor := 10
> 	tableID := sqlutils.QueryTableID(t, db, "d", "t")
> 	newDefCfg := *protoutil.Clone(s.(*server.TestServer).Cfg.DefaultZoneConfig).(*config.ZoneConfig)
> ```

Why does this need to be a clone?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao and @nvanbenschoten)

___
*[pkg/config/system.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXNtPj9wndemJsZ9VO:-LeXsb1n70KIrUDcVbPD:b-dkd0ds) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfea24909bdebf1fb9462f2c06b3e6a9d819c53/pkg/config/system.go#L74)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

ZoneConfig is 144 bytes large, so I think it's probably a good idea not to pass it around by value. If only we had `const &` in go.
</blockquote></details>

Well wait, I agree that on common call paths we should pass a zone config by reference but within this struct I don't see why it should be a pointer. SystemConfig is generally always already passed by reference. Having a nilable DefaultZoneConfig makes a client have to ask what a nil value means and whether this field is mutable by somebody else. Downstream I think it's reasonable and preferable to take the reference of this field. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @jeffrey-xiao)

___
*[pkg/config/system.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXNtPj9wndemJsZ9VO:-LeXtlwJ7h8GK2DNlZjv:b-76q79z) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfea24909bdebf1fb9462f2c06b3e6a9d819c53/pkg/config/system.go#L74)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Well wait, I agree that on common call paths we should pass a zone config by reference but within this struct I don't see why it should be a pointer. SystemConfig is generally always already passed by reference. Having a nilable DefaultZoneConfig makes a client have to ask what a nil value means and whether this field is mutable by somebody else. Downstream I think it's reasonable and preferable to take the reference of this field. 
</blockquote></details>

If we pass `SystemConfig` around by reference everywhere then I agree.


<!-- Sent from Reviewable.io -->

jeffrey-xiao commented:
___
*[pkg/config/zone.go, line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXQttJ0aucOOlbCFb5:-LeXtQiB9oVyzg5YLqP0:bes8fwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/config/zone.go#L271)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

@jeffrey-xiao and I talked about this and we decided to fix it in the same way in a follow-up PR. That way we can be sure we all agree on the approach for this one first. 
</blockquote></details>

I looked into the needed changes for defaultSystemZoneConfig and it seems like it's a fairly small change compared to defaultZoneConfig. I can probably add it to the PR if it doesn't add too much complexity.


<!-- Sent from Reviewable.io -->

jeffrey-xiao commented:
___
*[pkg/sql/show_trace_replica_test.go, line 44 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXq3B5-9iStjYfK8Xy:-LeY0s8LBvyrhtNG6U7w:b9wnfuv) ([raw file](https://github.com/cockroachdb/cockroach/blob/9485c3d76adcb0a6a0f1ef1750b8ef0ba1ca2e63/pkg/sql/show_trace_replica_test.go#L44)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We're able to get rid of this one already?
</blockquote></details>

The `TestingSetDefaultSystemZoneConfig` call should have been removed originally, but since I'm adding the changes for overriding DefaultSystemZoneConfig, it should be fine now.


<!-- Sent from Reviewable.io -->

jeffrey-xiao commented:
___
*[pkg/sql/zone_test.go, line 268 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXqLRT7uuuiLyW5h9f:-LeY2cHr5bZNb4-kLSPl:b-rkq0nw) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/sql/zone_test.go#L268)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why does this need to be a clone?
</blockquote></details>

I don't think it's needed. Removed.


<!-- Sent from Reviewable.io -->

jeffrey-xiao commented:
___
*[pkg/sql/set_zone_config.go, line 314 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXpxED0EJyuDg8u1k2:-LeXx8yB46LASabQ288U:bbxb23x) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/sql/set_zone_config.go#L314)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why do we need the clone here and below?
</blockquote></details>

completeZone and finalZone are modified later on.


<!-- Sent from Reviewable.io -->

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/config/system.go, line 74 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXNtPj9wndemJsZ9VO:-LeY00KK4jFQp9xSRk9G:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2dfea24909bdebf1fb9462f2c06b3e6a9d819c53/pkg/config/system.go#L74)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we pass `SystemConfig` around by reference everywhere then I agree.
</blockquote></details>

Done.
___
*[pkg/config/zone.go, line 336 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXQc2f-U2tx0MGVQZQ:-LeXyQT80L3o4aORBlnd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/config/zone.go#L336)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Make this 
```
zc := DefaultZoneConfig()
return &zc
```
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> Reviewed 37 of 60 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jeffrey-xiao and @nvanbenschoten)

___
*[pkg/config/zone.go, line 271 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeXQttJ0aucOOlbCFb5:-LeYHEdh0PMHGUFRJKWM:b-sa8zzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/983654a0a55cb145a4fcd5efe7491dadc38643d8/pkg/config/zone.go#L271)):*
<details><summary><i>Previously, jeffrey-xiao (Jeffrey Xiao) wrote…</i></summary><blockquote>

I looked into the needed changes for defaultSystemZoneConfig and it seems like it's a fairly small change compared to defaultZoneConfig. I can probably add it to the PR if it doesn't add too much complexity.
</blockquote></details>

Much appreciated!
___
*[pkg/sql/planner.go, line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeYIq0e7ENig7ZyYkvw:-LeYIq0e7ENig7ZyYkvx:b-j9gvdw) ([raw file](https://github.com/cockroachdb/cockroach/blob/172dd89675c9b5674ef20244aa0dc37c4d4707ba/pkg/sql/planner.go#L231)):*
> ```Go
> 		},
> 	}
> 	systemCfg := config.NewSystemConfig(execCfg.DefaultZoneConfig)
> ```

Well, this here is sort of a bummer. These are allocated for every statement if I understand correctly. I guess there's some value is saving some hundreds of bytes per planner. I don't know. I'm not sure I totally understand why we need to create a new SystemConfig for every planner. 

After digging through this code a bit it seems like this databaseCache doesn't really matter for the internal planner and the data in it never gets relied upon. Let's try stripping it out entirely (i.e. don't initialize the database cache on the tables) and see where we are. If it breaks things (which I think it may) then we can either deal with the slight inefficiency here (these internal planners are not used often) or we can pass an impl of the database cache that doesn't do anything. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> This is a really nice change. Thanks for taking the time to track down all the references. Let's see if @nvanbenschoten has other thoughts.

Should we create an issue or plan a follow-up PR to deal with the TODO on the databaseCache in the TableCollection? I don't know enough about the state of caching and leases in the SQL layer to make a determination.

> Reviewed 23 of 60 files at r3, 7 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-:-LeYj3vX8lWmZQ3_cNwY:brzhmen)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jeffrey-xiao and @nvanbenschoten)

___
*[pkg/sql/planner.go, line 224 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeYimlD3wk3BSeQ42fW:-LeYimlD3wk3BSeQ42fX:b1h4pk2) ([raw file](https://github.com/cockroachdb/cockroach/blob/bba74b1fdfeb28928e90867fa8f7a521947da430/pkg/sql/planner.go#L224)):*
> ```Go
> 	tables := &TableCollection{
> 		leaseMgr:      execCfg.LeaseManager,
> 		databaseCache: nil,
> ```

The zero value is nil so you just omit this. Maybe it's worth a comment that this table collection isn't hooked up to all the caches and system config the way a normal one would be. 


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> this is a great change!

My only remaining remark is that I don't think we should be copying these `ZoneConfig`s by value quite as much as we are (which may be rescinding on what I agreed to earlier). My expectation here would be that the pointer vs. value decision would reflect ownership. The owner of these default zone configurations is `server.Config`, so that should absolutely hold the structs by value. All other instances of the struct (e.g. in `ExecutorConfig`, `StoreConfig`, `SystemConfig`) are carrying around a server-level configuration. If these configs hold the default zone configs by value then it implies to me that they own the configuration and that the default zone configs shares the same scope as these objects, but that's not true. If these configs held the default zone config by reference then it would be clear that these are injected/borrowed from some outer scope.

The fact that struct is immutable makes this discussion a little more nuanced. Considering what would happen if the struct wasn't immutable clarifies things for me. We would need some internal synchronization and a clear ownership model. Holding these configs by value in the `server.Config` and passing references to them everywhere else gives us this ownership model.

> Having a nilable DefaultZoneConfig makes a client have to ask what a nil value means and whether this field is mutable by somebody else.

I'm not sure I agree with this in all cases. If the DefaultZoneConfig is injected in the constructor of `SystemConfig` by reference then I don't think there's much confusion about what a nil value means. Of course, it would certainly be nice if Go's type system allowed us to express the difference between `Option<Box<ZoneConfig>>` and `Box<ZoneConfig>`.

> Reviewed 27 of 60 files at r3, 5 of 7 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-:-LeZ0j259EaCGLaJD6Nj:buduyql)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner and @jeffrey-xiao)

___
*[pkg/sql/planner.go, line 224 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeYimlD3wk3BSeQ42fW:-LeZ06by5rXtZrfSGUNu:bjzx6uh) ([raw file](https://github.com/cockroachdb/cockroach/blob/bba74b1fdfeb28928e90867fa8f7a521947da430/pkg/sql/planner.go#L224)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

The zero value is nil so you just omit this. Maybe it's worth a comment that this table collection isn't hooked up to all the caches and system config the way a normal one would be. 
</blockquote></details>

Yeah, that is worth a comment.


<!-- Sent from Reviewable.io -->

ajwerner commented:
@nvanbenschoten is there a call where the ZoneConfig is passed by value that concerns you? When I looked through the PR the only places I saw it passed by value that wasn’t cluster bootstrap or server initialization was that newInternalPlanner which we’ve now eliminated. All the rest of the copies are in test setup AFAICT.

nvanbenschoten commented:
Not a specific call, but I fear that the config structs are getting severely bloated by having them all hold ZoneConfigs by value. Again, this is mostly a matter of making ownership clear right now because there aren't that many of these config structs, but imagine if we added a `ReplicaConfig`. Would we want that to hold one of these default ZoneConfigs by value?

ajwerner commented:
It sounds like we have similar goals but a different outlook. I agree that an ownership model is critical. That being said, the clearest way to present ownership is to put values into the configs. I agree that it’d be a problem to pass these things by value if we did it a non-constant number of times. I question the notion that config bloat is a problem. Configs of non-trivial size should only be created a constant number of times per package and generally define a clear ownership boundary, if you pass something a dependency in a config, that thing owns the dependency. If the owner dependency is mutable or stateful then it needs synchronization. I’m worried we’re introducing complexity as a micro-optimization to save a couple hundred bytes. I wouldn’t mind if config structs got to be >64kb, there are ~10 of them floating around. 

> Would we want that to hold one of these default ZoneConfigs by value?

That feels somewhat contrived. Replicas are owned by a store. There are a constant number of stores. If you want configs for objects which come and go throughout the life of a server, make a factory or something of that ilk. If we did have big replica configs that this was a part of then I’d be on your side or realistically fighting to get rid of those. 

All that said, a pointer with a well specified ownership model will work fine. My feelings aren’t as strong as maybe they sounded, just noting the premature optimization.

nvanbenschoten commented:
> I agree that an ownership model is critical. That being said, the clearest way to present ownership is to put values into the configs.

That's kind of the main point I'm trying to make. Storing these zone configs in objects like `ExecutorConfig`, `StoreConfig`, and `SystemConfig` makes it look like they own the default zone configs.. but they don't. The issue we're fixing here is to make zone configs a "property of a server" and I think the clearest way to represent that it to have the `Server` object hold the defaults by value and all other objects reference the `Server`'s value.

> That feels somewhat contrived.

Oh, it's very contrived 😃

jeffrey-xiao reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner and @nvanbenschoten)

___
*[pkg/sql/planner.go, line 231 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeYIq0e7ENig7ZyYkvw:-LelHFZK8otLIGz035J_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/172dd89675c9b5674ef20244aa0dc37c4d4707ba/pkg/sql/planner.go#L231)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Well, this here is sort of a bummer. These are allocated for every statement if I understand correctly. I guess there's some value is saving some hundreds of bytes per planner. I don't know. I'm not sure I totally understand why we need to create a new SystemConfig for every planner. 

After digging through this code a bit it seems like this databaseCache doesn't really matter for the internal planner and the data in it never gets relied upon. Let's try stripping it out entirely (i.e. don't initialize the database cache on the tables) and see where we are. If it breaks things (which I think it may) then we can either deal with the slight inefficiency here (these internal planners are not used often) or we can pass an impl of the database cache that doesn't do anything. 
</blockquote></details>

Done.
___
*[pkg/sql/planner.go, line 224 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-LeYimlD3wk3BSeQ42fW:-LelHF9HDz5IUcHnOzZI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bba74b1fdfeb28928e90867fa8f7a521947da430/pkg/sql/planner.go#L224)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Yeah, that is worth a comment.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
@ajwerner and I just came to an agreement that the `ZoneConfig` should be stored by value on the `ServerConfig` and a reference to that value should be stored everywhere else. For instance, when passing the default zone config down to the `ExecutorConfig`, we'd do something like:
```
ExecutorConfig{
    DefaultZoneConfig: &serverCfg.DefaultZoneConfig,
}
```

How does that sound?

jeffrey-xiao committed:
server: add default zone into server config, store config, executor config, systemconfig, and gossip

Previously the default zone config was stored as a global and code in
various layers would use `config.DefaultZoneConfig()` directly. It was
difficult for parallel tests to modify the global and restore it
correctly.

This commit adds the default zone config to the server config, store
config, executor config, system config, and gossip so it is accessible
in the places where it's used.

The default zone config can be overridden in tests by setting the
`DefaultZoneConfigOverride` testing knob.

Resolves: #37054

Release note: None

jeffrey-xiao committed:
server: initialize default zone config and update functions and tests to use it

This commit populates the default zone config in the server config,
store config, executor config, system config, and gossip. Additionally,
all usages of the global default zone config are replaced with the
default zone configs stored in the mentioned places.

Resolves: #37054

Release note: None

jeffrey-xiao commented:
> @ajwerner and I just came to an agreement that the `ZoneConfig` should be stored by value on the `ServerConfig` and a reference to that value should be stored everywhere else. For instance, when passing the default zone config down to the `ExecutorConfig`, we'd do something like:
> 
> ```
> ExecutorConfig{
>     DefaultZoneConfig: &serverCfg.DefaultZoneConfig,
> }
> ```
> 
> How does that sound?

Sounds good! I've made the changes to reflect this.

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for bearing with us

> Reviewed 3 of 67 files at r5, 64 of 64 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-:-Les1L2k92kfKQj8tIhO:b-6mramx)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice job on this!

To merge, you just have to comment something like:

bors r=ajwerner,nvanbenschoten

See https://cockroachlabs.atlassian.net/wiki/spaces/CRDB/pages/73204099/Bors+Merge+Bot for reference.

> Reviewed 1 of 67 files at r5, 32 of 64 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37434#-:-Lesa_6x8cSYnV0lCiEe:bhcvmg)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->
